src/PageParser.php

```diff
@@ -108,32 +107,37 @@ public function getChaptersList( $pageList, $namespaces ) {
 	}
 
 	/**
-	 * return the list of the chapters with the summary if it exist, if not find links to subpages.
+	 * Get a list of the chapters in the given work,
+	 * using the ws-summary if it exists and all linked subpages if it doesn't.
+	 * @param string $title The page title.
+	 * @param array<int,string> $pageList
+	 * @param array<int,string> $namespaces
 	 * @return Page[]
 	 */
-	public function getFullChaptersList( $title, $pageList, $namespaces ) {
+	public function getFullChaptersList( string $title, array $pageList, array $namespaces ) {
 		$chapters = $this->getChaptersList( $pageList, $namespaces );
 		if ( empty( $chapters ) ) {
-			$list = $this->xPath->query( '//a[contains(@href,"/' . Util::wfUrlencode( $title ) . '/") and
+			// If no chapters were found in the usual way, try to guess the chapters
+			// by finding all links that start with the current page's name.
+			$list = $this->xPath->query( '//a[
 				not(
 					contains(@class,"new") or
 					contains(@class,"extiw") or
 					@rel = "mw:WikiLink/Interwiki" or
 					contains(@class,"external") or
-					contains(@href,"#") or
```

Comment:
it would be nice to have tests to check what happen with fragment links to the same page and to a subpage. I don't remind what is the Parsoid output for them. ws-export had troubles with them in the past.

Reply:
Good idea. Done. It doesn't clash because in this patch I'm adding the prefix check with the trailing slash, which will never happen for a self-link.